server/ipqueue.go

```diff
@@ -19,27 +19,34 @@ import (
 	"sync/atomic"
 )
 
-const ipQueueDefaultMaxRecycleSize = 4 * 1024
+const (
+	ipQueueDefaultMaxRecycleSize = 4 * 1024
+	ipQueueMaxPoolSize           = 5
+)
 
 // This is a generic intra-process queue.
+// Fields are in specific order based on benchmark results.
 type ipQueue[T any] struct {
-	inprogress int64
+	inprogress   int64
+	inprogressSz uint64
 	sync.Mutex
+	ipQueueOpts[T]
 	ch   chan struct{}
 	elts []T
 	pos  int
-	pool *sync.Pool
```

Comment:
This makes me a bit uncomfortable, as the important secondary property of a `sync.Pool` is that they can be automatically emptied by the GC, which isn't true of our own `pool` implementation (or at the very least, not until we get weak pointer support in Go ~1.24).

As I see it, there are two possible solutions to the SA6002 problem with `sync.Pool` and slices. The first is to either use a pointer to the slice header and reuse it, with the understanding that we will need to keep an eye on the underlying capacity to ensure we don't re-pool massive arrays (the max recycle size). The second is to use a pointer to a fixed-length array and slice it on-demand (like how the `client.go` buffers now do). I don't suppose either will really make a huge difference in this case because the value from the pool will escape to the heap anyway by storing it in the `ipQueue` struct. 

But all else equal, I think it's a far better idea to stick with `sync.Pool` for the GC properties alone.

Reply:
Ok, then I can get back to the first version of this PR (unfortunately, I pushed-force the changes on it) to use a `*[]T` and go back to a sync.pool. For that to work, we need the pop() to return the *[]T and pass that to recycle(), otherwise, we would have the problem described in the PR description.